modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C

```diff
@@ -220,14 +220,35 @@ SingleVariableReturnMappingSolution::internalSolve(const Real effective_trial_st
         {
           if (scalar_upper_bound != max_permissible_scalar && scalar_lower_bound != 0.0)
           {
-            Real frac = 0.5;
-            scalar_increment =
-                (1.0 - frac) * scalar_lower_bound + frac * scalar_upper_bound - scalar_old;
-            modified_increment = true;
-            if (iter_output)
-              *iter_output << "  Trial scalar_increment exceeded bounds.  Setting between "
-                              "lower/upper bounds. frac: "
-                           << frac << std::endl;
+            const Real tol = 1e-14;
+            // Check to avoid divide by zero or overflow in calculation of scalar_increment
+            if (std::abs(1.0 - residual_old / residual) > tol)
+            {
+              scalar_increment =
+                  residual * (scalar_old - scalar) / (residual - residual_old) + scalar;
+              if (scalar_old + scalar_increment < scalar_upper_bound &&
+                  scalar_old + scalar_increment > scalar_lower_bound)
+              {
+                modified_increment = true;
+                if (iter_output)
+                  *iter_output << "  Trial scalar_increment exceeded bounds.  Setting between "
```

Comment:
I don't understand the logic here. _If less than upper bound and greater than lower bound_ then output _exceeds bounds_? Looks like it is exactly _within_ bounds here. Maybe add comments?

Reply:
If outside bounds, we need to adjust.  If the secant worked (now in bounds), note that we've `modified_increment` output a message saying that we've adjusted.  If the secant didn't work (still out of bounds), fall back on bisect and output a message that we've adjusted.